### PR TITLE
Update discovery nodes after cluster state is published

### PR DESCRIPTION
**Pull request by abeyad**
Before, when there was a new cluster state to publish,
zen discovery would first update the set of nodes to
ping based on the new cluster state, then publish the new
cluster state. This is problematic because if the cluster
state failed to publish, then the set of nodes to ping
should not have been updated.

This commit fixes the issue by updating the set of
nodes to ping for fault detection only _after_ the new
cluster state has been published.

### CODE REVIEW

**Comment by abeyad** [comment]
@bleskes FYI

**Comment by s1monw** [comment]
LGTM

File: core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryUnitTests.java [hunk 1]
```diff
@@ -127,4 +146,78 @@ public void testFilterNonMasterPingResponse() {
...
+                // publishing a new cluster state
+                ClusterChangedEvent clusterChangedEvent = new ClusterChangedEvent("testing", newState, state);
+                AssertingAckListener listener = new AssertingAckListener(newState.nodes().getSize() - 1);
+                expectedFDNodes = zenDiscovery.getFaultDetectionNodes();
+                zenDiscovery.publish(clusterChangedEvent, listener);
+                listener.await(1, TimeUnit.HOURS);
+                // publish was a success, update expected FD nodes based on new cluster state
+                expectedFDNodes = fdNodesForState(newState, master.discoveryNode);
+            } catch (Discovery.FailedToCommitClusterStateException e) {
+                // not successful, so expectedFDNodes above should remain what it was originally assigned
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryUnitTests.java]: can we capture whether we set min master node to 3 or 1 and check here that the exception is really expected?

**Comment by bleskes** [comment]
LGTM2. Though it will be good to see CI agree

**Comment by bleskes** [comment]
retest this please

**Comment by abeyad** [comment]
retest this please

File: core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryUnitTests.java (hunk 1 above)
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryUnitTests.java]: I added this in 72b8711bd70c04fa1ca8b362e2c3e952d5d041c4

**Comment by abeyad** [comment]
CI keeps failing because of https://github.com/elastic/infra/issues/1372
Trying again..

**Comment by abeyad** [comment]
retest this please

**Comment by abeyad** [comment]
All checks are clear, thanks for the review @s1monw and @bleskes

**Comment by abeyad** [comment]
5.0 commit: ba27a04c102fbbcde0420ac03c70fa3ea98cb76b
5.x commit: 90749eb14d0fd903a7b1d31df6eeb7aaadd6dd85

**Pull request merged**